### PR TITLE
Backport 3.6: md: allow dispatch to PSA whenever CRYPTO_CLIENT is enabled

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

Instead of allowing PSA dispatching only when CRYPTO_C is set and some MBEDTLS_PSA_ACCEL_ALG_xxx is set, we enable dispatching when CRYPTO_CLIENT and PSA_WANT_ALG_xxx are set. This makes the feature more useful in cases where the PSA support is provided externally, like for example TF-M in Zephyr.

## PR checklist

- [x] **changelog** provided
- [x] **development PR** not required because crypto only
- [x] **crypto PR** https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/188
- [x] **framework PR** https://github.com/Mbed-TLS/mbedtls-framework/pull/140
- [x] **3.6 PR** here
- [x] **2.28 PR** not required because new feature
- [x] **tests**  not required because this only affects client-server builds which are only unofficially supported and a known test gap in 3.6

### CODE REVIEW

**Comment by valeriosetti** [comment]
@mpg I took the liberty to add you as reviewer because you were the one who improved MD last year, allowing for the PSA dispatch, and we also discussed about this change in Slack. However if you don't have enough review bandwidth for this, please let me know and I'll remove you ;)

**Comment by valeriosetti** [comment]
The ABI-API break is expected because `struct mbedtls_md_context_t` will change its size depending on `PSA_WANT_` instead of `MBEDTLS_PSA_ACCEL_` symbols.

**Comment by mpg** [comment]
> * **changelog** TBD, but I would like to have some feedback from reviewers about this change before doing it.

IMO not ChangeLog for this, as support for `CLIENT && !C` is not official yet, and when it will become official it will have a ChangeLog entry with a scope much broader than just this :)

**Comment by mpg** [comment]
> * **development PR** not required because, if I'm not wrong, the MD module is internal in `development`, so the user is not able to make direct usage of it.

The current plan is for MD to remain public in 4.0, so I think we want to forward-port this to development. However there might be other changes planned (or already done) in development that would make this moot, I'm not sure. Cc @gilles-peskine-arm

**Comment by mpg** [comment]
> The ABI-API break is expected because `struct mbedtls_md_context_t` will change its size depending on `PSA_WANT_` instead of `MBEDTLS_PSA_ACCEL_` symbols.

Aw, I'm afraid this is a blocker. We [promise not to change the ABI in LTS branches](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#long-time-support-branches) (which 3.6 now is) unless we can't find another way to fix a security issue. So, changing the size of `struct mbedtls_md_context_t` in the default configuration is not something we can do in 3.6.

Would there be a way to achieve what you're after while keeping the structure unchanged in the default config, and only changing it in a few specific configs?

**Review by mpg — CHANGES_REQUESTED** [review]
https://github.com/Mbed-TLS/mbedtls/pull/9562#issuecomment-2354827277

**Comment by valeriosetti** [comment]
> The current plan is for MD to remain public in 4.0, so I think we want to forward-port this to development.

Oh, I see. I based my previous comment on the fact that `md.h` header is now in `tf-psa-crypto/drivers/builtin/include/mbedtls/md.h` which looks like a quite "hidden" place not meant to be included by end user's code directly. However I surely missed something, so please apologize for the wrong assumption.

> Would there be a way to achieve what you're after while keeping the structure unchanged in the default config, and only changing it in a few specific configs?

I think the easiest way will be to enable PSA dispatching based on PSA_WANTs only in the usual `CRYPTO_CLIENT && !CRYPTO_C` configuration. This is still relevant for Mbed TLS users like Zephyr, but it does not affect the default configuration. I will try to rework the PR in this direction.

**Comment by gilles-peskine-arm** [comment]
> > The current plan is for MD to remain public in 4.0, so I think we want to forward-port this to development.
>
> Oh, I see. I based my previous comment on the fact that md.h header is now in tf-psa-crypto/drivers/builtin/include/mbedtls/md.h which looks like a quite "hidden" place not meant to be included by end user's code directly. However I surely missed something, so please apologize for the wrong assumption.

Currently all of the legacy crypto headers have moved as part of the work to split the repositories. We'll move some of them again as part of the work to evolve the API. [`md.h` will remain public](https://github.com/Mbed-TLS/mbedtls/issues/8450) (but without HMAC functionality). [Most other crypto headers will become private or unstable](https://github.com/Mbed-TLS/mbedtls/issues/8663).

**Comment by valeriosetti** [comment]
> > Would there be a way to achieve what you're after while keeping the structure unchanged in the default config, and only changing it in a few specific configs?
> 
> I think the easiest way will be to enable PSA dispatching based on PSA_WANTs only in the usual `CRYPTO_CLIENT && !CRYPTO_C` configuration. This is still relevant for Mbed TLS users like Zephyr, but it does not affect the default configuration. I will try to rework the PR in this direction.

Thanks to this change the ABI-API failure disappeared and the CI is fully green now :)

**Review by mpg — APPROVED** [review]
LGTM, thanks!

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I'm ok with this change in principle. And I agree that since this is targeting a configuration that is not officially supported (`MBEDTLS_PSA_CRYPTO_CLIENT` without `MBEDTLS_PSA_CRYPTO_C`), we should in principle not add a changelog entry. But still, I want more documentation for two audiences:

* For maintainers, there should be an explanation of when md dispatches to PSA in plain English (as opposed to C code spread over multiple files). This is in principle covered by `docs/architecture/psa-migration/md-cipher-dispatch.md`, though I'm not sure it was fully up to date even before this pull request. Please update that document. And I'd like a summary, or a reference to that file, at least in `md.c`.
* For users who use CLIENT without C (and who enable at least one hash), there is an incompatible change: they now need to implement `psa_can_do_hash`. We should document that somewhere, even if we keep saying that it's experimental. I guess in the documentation of `MBEDTLS_PSA_CRYPTO_CLIENT`?

Note that I'm not convinced that we shouldn't add a changelog entry. This is a long-time support branch, so the threshold for considering a change to be important it low. We do sometimes announce things like “The undocumented ability … has been removed” in the changelog. I think now is one of those times.

**Comment by mpg** [comment]
Ok, I'm not opposed to a ChangeLog entry as long as it reminds readers that `CLIENT && !C` is still not officially supported.

Good point about updating the documentation, thanks for noticing!

**Comment by valeriosetti** [comment]
The last force-push was mostly to reset the head of this PR on top of the current `mbedtls-3.6` because I have not been working on this PR for a long time and it was quite outdated.

**Comment by mpg** [comment]
Note that the CI is failing in `test_default_psa_crypto_client_without_crypto_provider` which looks related.

**Comment by valeriosetti** [comment]
> Note that the CI is failing in `test_default_psa_crypto_client_without_crypto_provider` which looks related.

Yes, this is due to a change that I recently introduced https://github.com/Mbed-TLS/mbedtls/pull/9562/commits/cf8bbb88043b0fde4911ab561c65b8677dcc24f3. To me that made sense, but apparently this is triggering a lot of failures in that test scenarios because the library code "thinks" that the PSA Crypto supports hash functions, whereas they are all stub functions. I need to think a bit more on how to overcome this problem. In the worst case I'll just drop the commit.

**Comment by valeriosetti** [comment]
> > Note that the CI is failing in `test_default_psa_crypto_client_without_crypto_provider` which looks related.
> 
> Yes, this is due to a change that I recently introduced [cf8bbb8](https://github.com/Mbed-TLS/mbedtls/commit/cf8bbb88043b0fde4911ab561c65b8677dcc24f3). To me that made sense, but apparently this is triggering a lot of failures in that test scenarios because the library code "thinks" that the PSA Crypto supports hash functions, whereas they are all stub functions. I need to think a bit more on how to overcome this problem. In the worst case I'll just drop the commit.

I think I've found a workaround for the problem I mentioned above -> https://github.com/Mbed-TLS/mbedtls/compare/304bd4f3efd424f762455b226a7afa28bbdcef12..5368386524f9432fb21179e7800dfdb4c861c824.
Wdyt?

**Comment by valeriosetti** [comment]
@mpg @gilles-peskine-arm the CI is green again, so the PR should be ready for another round of reviews when/if you have some spare time :)

**Comment by frkv** [comment]
We deem this essential to ensure Zephyr supports TF-M out-of-the-box for TLS/DTLS and X.509 use-cases using the LTS branch (of TF-M and Mbed TLS)

Zephyr actively uses `MBEDTLS_PSA_CRYPTO_CLIENT` 

Please prioritize this for Mbed TLS 3.6.3. This seems like a very low hanging fruit for better PSA crypto adoption (and by extension essential for TF-M support)

File: library/md.c [hunk 1]
```diff
@@ -248,7 +248,24 @@ static int md_can_use_psa(const mbedtls_md_info_t *info)
...
+     * case we try to simply abort a freshly created hash operation. If it fails
+     * we assume that either psa_crypto_init() has not been called or the PSA
+     * Crypto provider does not support hash.
+     */
+#if defined(MBEDTLS_PSA_CRYPTO_C)
     return psa_can_do_hash(alg);
+#else /* MBEDTLS_PSA_CRYPTO_CLIENT */
+    psa_hash_operation_t operation = PSA_HASH_OPERATION_INIT;
+    psa_status_t status = psa_hash_abort(&operation);
+    if ((status == PSA_ERROR_COMMUNICATION_FAILURE) || (status == PSA_ERROR_NOT_SUPPORTED)) {
```
Comment (mpg) [inline library/md.c:264]: Nice trick. However wouldn't it be safer to return 0 on _any_ error? Checking [the documentation](https://arm-software.github.io/psa-api/crypto/1.1/api/ops/hashes.html#c.psa_hash_abort) it can also return `PSA_ERROR_BAD_STATE`, and I think drivers are in theory free to return other error codes.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks pretty good to me, just a couple comments.

File: library/md.c [hunk 2]
```diff
@@ -248,7 +248,24 @@ static int md_can_use_psa(const mbedtls_md_info_t *info)
...
+    /* psa_can_do_hash() below is used to determine if the PSA Crypto core has
+     * been initialized or not. However it is a vendor specific function which
+     * is not part of the standard PSA Crypto interface so we cannot make any
+     * assumption about it being supported in the PSA Crypto provider. In this
+     * case we try to simply abort a freshly created hash operation. If it fails
+     * we assume that either psa_crypto_init() has not been called or the PSA
+     * Crypto provider does not support hash.
+     */
+#if defined(MBEDTLS_PSA_CRYPTO_C)
     return psa_can_do_hash(alg);
```
Comment (gilles-peskine-arm) [inline library/md.c:260]: Shouldn't we call `psa_can_do_hash` in the client-only case as well? I don't see why we wouldn't.

File: library/md.c (hunk 1 above)
Comment (gilles-peskine-arm) [reply library/md.c:264]: I'm also not sure whether this is the right set of errors, but it would make sense to treat most error codes as final. We're asking a subsystem “can you do this?” There are four kinds of answers: “yes”, “not at this time but maybe later”, “not me but maybe someone else” and “this is impossible”. Since we don't cache the result or retry in a loop, we don't care about the difference between “not at this time” and “not me”. On the other hand, “I can't” and “nobody can” are different conditions: for “I can't”, it is desirable to try another method to perform the same task; for “nobody can”, we should give up immediately.

In the driver interface, “not me” is `PSA_ERROR_NOT_SUPPORTED` (leading to trying a fallback driver if there is one) and “this is impossible” is all other error codes. (Again, there is no “not now” because there is no retry or caching mechanism.) Only `NOT_SUPPORTED` is treated as “not me” because this is what the fallback mechanism is meant for: when a driver only supports a subset of possible inputs, and there might be another driver that is (for example) slower but more featured. Other errors such as `INVALID_ARGUMENT` are fatal because they assert that the driver knows what it's doing, and it knows that the request is not valid.

Here, I think `COMMUNICATION_FAILURE` makes sense as a “not now” error: it's the most natural error code if there's a client-server mechanism but it is not available yet during system startup. (For example, in a bootloader which loads signed drivers, and needs to calculate hashes to verify the driver's signatures.) I would expect most other errors to be fatal. `BAD_STATE` would also be a reasonable error code in that case; it's a “not like this” error, and for a freshly initialized operation, the only part of “like this” that could be invalid is the system state. In particular, `NOT_SUPPORTED` means that `psa_hash_abort` is not supported at all, so whatever PSA implementation we have is not at all capable of calculating hashes. But if that's the case, then we shouldn't reach this point: all `PSA_WANT_ALG_hash` should be disabled and so `psa_alg_of_md` should have returned 0.

The set of errors that we have here is part of the contract between Mbed TLS and the PSA Crypto provider. So we should document it somewhere, and it should follow some simple rule. Since this is not an official API, we can change the rule (but we'll try not to change it without a good reason). I'm not sure what the right place is for this documentation.

All this being written: do we really need this check? My inclination would be to let the integrator worry about this. Require the integrator to implement `psa_can_do_hash`, and just use its boolean result.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Having analyzed the design more closely (with the help of the added documentation, thanks for that), I think it's geared towards one specific integration, and putting the availability check in the wrong place. Whether a hash is available through PSA at a given time is very integration-dependent, so I think that should be the job of `psa_can_do_hash`, to be provided by the integrator in client-only builds.

So I am rejecting a design that bypasses `psa_can_do_hash` in client-only builds, unless you give a good reason why we should do it this way.

Once we converge on the design, please make a forward-port to the development branch. It doesn't matter if development isn't in its final state, just make a similar patch as if `md.h` was going to be released as is, and we'll continue from there later.

File: library/md.c (hunk 2 above)
Comment (valeriosetti) [reply library/md.c:260]: As discussed during the meeting `md_can_use_psa` in case of `MBEDTLS_PSA_CRYPTO_CLIENT && !MBEDTLS_PSA_CRYPTO_C` now:

- checks that the hash algorithm is supported in PSA through `psa_alg_of_md`
- checks that `psa_crypto_init` was called through the fake aborting of a freshly initialized hash operation.

Is that reasonably OK?

File: docs/architecture/psa-migration/md-cipher-dispatch.md [hunk 3]
```diff
@@ -290,7 +296,7 @@ We could go further and make PSA accelerators available to legacy callers that c
 
 It is strongly desirable to allow mechanisms available through PSA but not legacy: this allows saving code size when an accelerator is present.
 
-There is no strong reason to allow mechanisms available through legacy but not PSA when `MBEDTLS_PSA_CRYPTO_C` is enabled. This would only save at best a very small amount of code size in the PSA dispatch code. This may be more desirable when `MBEDTLS_PSA_CRYPTO_CLIENT` is enabled (having a mechanism available only locally and not in the crypto service), but we do not have an explicit request for this and it would be entirely reasonable to forbid it.
+There is no strong reason to allow mechanisms available through legacy but not PSA when `MBEDTLS_PSA_CRYPTO_C` is enabled. This would only save at best a very small amount of code size in the PSA dispatch code.
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/md-cipher-dispatch.md:299]: Why did you remove the part about CLIENT here? It's still relevant.

File: docs/architecture/psa-migration/md-cipher-dispatch.md [hunk 4]
```diff
@@ -190,8 +197,7 @@ Note: PSA cipher is built on Cipher, but PSA AEAD directly calls the underlying
 
 Here are some reasons why calling `psa_xxx()` to perform a hash or cipher calculation might not be desirable in some circumstances, explaining why the application would arrange to call the legacy software implementation instead.
 
-* `MBEDTLS_PSA_CRYPTO_C` is disabled.
-* There is a PSA driver which has not been initialized (this happens in `psa_crypto_init()`).
+* For ciphers `MBEDTLS_PSA_CRYPTO_CLIENT` is disabled.
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/md-cipher-dispatch.md:200]: This isn't specific to ciphers. If `MBEDTLS_PSA_CRYPTO_CLIENT` is disabled, then there are no `psa_xxx` functions in the build, so we can't call them.

Also, the part about drivers is still relevant.

```suggestion
* `MBEDTLS_PSA_CRYPTO_CLIENT` is disabled.
* There is a PSA driver which has not been initialized (this happens in `psa_crypto_init()`).
```

File: docs/architecture/psa-migration/md-cipher-dispatch.md [hunk 5]
```diff
@@ -488,15 +491,24 @@ static inline psa_algorithm_t psa_alg_of_md_info(
...
 
-As a starting point, return 1 if PSA crypto's driver subsystem has been initialized.
+* If `MBEDTLS_PSA_CRYPTO_C` is set `int psa_can_do_hash(psa_algorithm_t hash_alg)` is used to further checking if the PSA Crypto core has been initialized or not. If so then `md_can_use_psa` will finally succeed, otherwise it will fail.
 
-Usage note: for algorithms that are not enabled via PSA, calling `psa_can_do_hash` is generally safe: whether it returns 0 or 1, you can call a PSA hash function on the algorithm and it will return `PSA_ERROR_NOT_SUPPORTED`.
+* If `MBEDTLS_PSA_CRYPTO_CLIENT && !MBEDTLS_PSA_CRYPTO_C`, `psa_can_do_hash` cannot be called because this is an Mbed TLS vendor specific addition that is not officially supported by the PSA Crypto standard. Therefore in this case `md_can_use_psa` will simply try to abort a freshly initialized `psa_hash_operation_t`; if it fails `md_can_use_psa` will assume that `psa_crypto_init()` has not been called.
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/md-cipher-dispatch.md:511]: I don't understand why attempting to abort an operation would be relevant. This looks to me like an implementation-specific hack. Some plausible behaviors would be incompatible with this check, such as:

* The local code of abort sees that it has no handle of an ongoing operation, so does nothing without attempting to contact the server.
* The first attempt to perform an operation attempts to open a connection. This may or may not be desirable during the early boot where a fallback may be desirable.

The considerations for “has client-server communication been initialized yet?” and “have drivers [and the keystore, if we extend this to ciphers] been initialized yet?” are similar. For “have hash/cipher drivers been initialized yet?”, we introduced `psa_can_do_hash` and `psa_can_do_cipher`. Those functions are provided by our library when `MBEDTLS_PSA_CRYPTO_C` is enabled. When only `MBEDTLS_CRYPTO_CLIENT_C` is enabled, it would make sense to require the implementation of the communication layer to provide them.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I still disagree with the proposed implementation, which looks to me like a vendor-specific trick that would be unlikely to work with other vendors.

I propose to follow the architecture proposed in the section “Improving support for `MBEDTLS_PSA_CRYPTO_CLIENT`”, unless there's a better proposal.

File: docs/architecture/psa-migration/md-cipher-dispatch.md [hunk 6]
```diff
@@ -488,15 +492,22 @@ static inline psa_algorithm_t psa_alg_of_md_info(
...
 ```
 
-The job of this private function is to return 1 if `hash_alg` can be performed through PSA now, and 0 otherwise. It is only defined on algorithms that are enabled via PSA.
+Internally this function does the following:
 
-As a starting point, return 1 if PSA crypto's driver subsystem has been initialized.
+* First of all it converts the `mbedtls_md_info_t` to `psa_algorithm_t`. The result of this conversion is based on the `MBEDTLS_MD_xxx_VIA_PSA` symbols: if an algorithm does not have the corresponding `MBEDTLS_MD_xxx_VIA_PSA` enabled, then `md_can_use_psa` will return false.
 
-Usage note: for algorithms that are not enabled via PSA, calling `psa_can_do_hash` is generally safe: whether it returns 0 or 1, you can call a PSA hash function on the algorithm and it will return `PSA_ERROR_NOT_SUPPORTED`.
+* `int psa_can_do_hash(psa_algorithm_t hash_alg)` is then used to further checking if the PSA Crypto core has been initialized or not. If so then `md_can_use_psa` will finally succeed, otherwise it will fail.
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/md-cipher-dispatch.md:510]: Please add a note to explain that in client-server builds, the implementer of the client interface is expected to provide `psa_can_do_hash()`.

File: ChangeLog.d/9652.txt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+Features
+   * MD module can now perform PSA dispatching also when
+     `MBEDTLS_PSA_CRYPTO_CLIENT && !MBEDTLS_PSA_CRYPTO_C`, even though this
+     configuration is not officially supported. This requires that a
+     PSA Crypto provider library which support the required `PSA_WANT_ALG_xxx`
+     is linked against Mbed TLS and that `psa_crypto_init()` is called before
+     performing any PSA call.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/9652.txt:7]: Also the provider needs to implement `psa_can_do_hash()`. Please:

* Mention this in the changelog entry.
* Move the declaration of `psa_can_do_hash()` to a public header file, with some suitable documentation. In principle this should be a new header, to be consumed only by client providers. In practice I think we can use `include/psa/crypto_extra.h` in 3.6, in a new Doxygen group.

```
/** \defgroup psa_crypto_client Functions defined by a client provider
 *
 * The functions in this group are meant to be implemented by providers of
 * the PSA Crypto client interface. They are provided by the library when
 * #MBEDTLS_PSA_CRYPTO_C is enabled.
 *
 * \note All functions in this group are experimental, as using
 *       alternative client interface providers is experimental. 
 *
 * @{
 */
/** … */
psa_status_t psa_can_do_hash(…);
/**@}*/

```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM except that we should document `psa_can_do_hash()` since it's now an interface that client providers need to implement.

**Review by gilles-peskine-arm — DISMISSED** [review]
Thanks for updating the documentation! Looks good to me now.

**Review by mpg — APPROVED** [review]
LGTM, thanks!

**Comment by mpg** [comment]
@valeriosetti can you update the submodule pointer now that https://github.com/Mbed-TLS/mbedtls-framework/pull/140 has been merged?

**Review by mpg — APPROVED** [review]

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM after framework update

**Comment by gilles-peskine-arm** [comment]
Can you please make a PR to development? Sure, the md interface will change, but that doesn't affect the changes here.

**Comment by gilles-peskine-arm** [comment]
`md.h` will remain public in TF-PSA-Crypto 1.x, with a subset of functionality: just hash calculation, not HMAC or metadata. Basically the current `MBEDTLS_MD_LIGHT`.
